src/search/search_engines/hierarchical_search_engines/iterative_width_search.cc

```diff
@@ -0,0 +1,148 @@
+#include "iterative_width_search.h"
+
+#include "goal_test.h"
+
+#include "../../option_parser.h"
+#include "../../plugin.h"
+#include "../../tasks/root_task.h"
+#include "../../tasks/propositional_task.h"
+#include "../../task_utils/successor_generator.h"
+#include "../../task_utils/task_properties.h"
+#include "../../utils/logging.h"
+#include "../../utils/memory.h"
+
+#include <cassert>
+#include <cstdlib>
+
+using namespace std;
+using namespace hierarchical_search_engine;
+
+namespace iw_search {
+IWSearch::IWSearch(const Options &opts)
+    : HierarchicalSearchEngine(opts),
+      width(opts.get<int>("width")),
+      debug(opts.get<utils::Verbosity>("verbosity") == utils::Verbosity::DEBUG),
+      m_novelty_base(nullptr),
+      m_novelty_table(0) {
+}
+
+bool IWSearch::is_novel(const State &state) {
+    return m_novelty_table.insert(dlplan::novelty::TupleIndexGenerator(m_novelty_base, m_propositional_task->get_fact_ids(state)), true);
+}
+
+bool IWSearch::is_novel(const OperatorProxy &op, const State &succ_state) {
+    return m_novelty_table.insert(dlplan::novelty::TupleIndexGenerator(m_novelty_base, m_propositional_task->get_fact_ids(op, succ_state)), true);
+}
+
+void IWSearch::print_statistics() const {
+    statistics.print_detailed_statistics();
+    m_search_space->print_statistics();
+}
+
+SearchStatus IWSearch::step() {
+    if (open_list.empty()) {
+        utils::g_log << "Completely explored state space -- no solution!" << endl;
+        return FAILED;
+    }
+    StateID id = open_list.front();
+    open_list.pop_front();
+    State state = m_state_registry->lookup_state(id);
+    SearchNode node = m_search_space->get_node(state);
+    node.close();
+    assert(!node.is_dead_end());
+    statistics.inc_expanded();
+
+    /* Goal check in initial state. */
+    if (id == m_initial_state->get_id()) {
+        if (m_goal_test->is_goal(*m_initial_state, state)) {
+            return on_goal_leaf(state);
+        }
+    }
+
+    vector<OperatorID> applicable_ops;
+    successor_generator.generate_applicable_ops(state, applicable_ops);
+    for (OperatorID op_id : applicable_ops) {
+        OperatorProxy op = task_proxy.get_operators()[op_id];
+        if (node.get_real_g() + op.get_cost() >= bound) {
+            continue;
+        }
+
+        int old_num_states = m_state_registry->size();
+        State succ_state = m_state_registry->get_successor_state(state, op);
+        int new_num_states = m_state_registry->size();
+        bool is_new_state = (new_num_states > old_num_states);
+        if (!is_new_state) {
+            continue;
+        }
+
+        statistics.inc_generated();
+        bool novel = is_novel(succ_state);
```

Comment:
Generalized implementation of NoveltyTable for arbitrary k in N is currently unable to consider operator effects that must occur in a tuple.